src/box/lua/iproto.c

```diff
@@ -268,16 +363,122 @@ push_iproto_protocol_features(struct lua_State *L)
 					  lengthof(features));
 }
 
+static void
+mpstream_error_cb(void *error_ctx)
+{
+	*(bool *)error_ctx = true;
+}
+
+/**
+ * Converts a packet part argument to MsgPack: if the argument is a string, then
+ * no conversion is needed — otherwise the argument must be a Lua table. The Lua
+ * table is encoded to MsgPack using IPROTO key translation table.
+ * In both cases, the result is stored on the fiber region.
+ */
+static const char *
+process_packet_part_arg(struct lua_State *L, int idx, size_t *mp_len)
+{
+	int packet_part_type = lua_type(L, idx);
+	if (packet_part_type != LUA_TSTRING && packet_part_type != LUA_TTABLE) {
+		lua_pushfstring(L, "expected table or string as %d argument",
+				idx);
+		return NULL;
+	}
+
+	struct region *gc = &fiber()->gc;
+	if (packet_part_type == LUA_TSTRING) {
+		const char *arg = lua_tolstring(L, idx, mp_len);
+		char *mp = xregion_alloc(gc, *mp_len);
+		return memcpy(mp, arg, *mp_len);
+	}
+
+	bool is_error = false;
+	struct mpstream stream;
+	mpstream_init(&stream, gc, region_reserve_cb, region_alloc_cb,
+		      mpstream_error_cb, &is_error);
```

Comment:
We usually use `luamp_error` in this case.

Reply:
Not always: https://github.com/tarantool/tarantool/blob/943f7d159caefaa28cc32ebc036f29eccb6325d4/src/box/xrow.c#L575-L579

The problem with `luamp_error` is that it throws a Lua exception, which we cannot catch, and which can cause the fiber region to leak.

Reply:
> Not always

By "in this case" I meant when there's a Lua state available. xrow/iproto is clearly not this case.

> The problem with luamp_error is that it throws a Lua exception, which we cannot catch, and which can cause the fiber region to leak.

We leak memory if the encoder fails anyway, which is much more likely than memory allocation error. I don't see any point in fixing one source of leaks, but leaving the other in the same function. We could probably wrap the whole function in `lua_cpcall`, but I don't know how it'd affect performance.

Reply:
Let's leave this up to @alyapunov.

Reply:
Fixed: changed the error handler to `luamp_error`.